File: Algorithm/2023.11/wolfy916/알고리즘 테마정리/다이나믹 프로그래밍/정수 삼각형.py
'''
정수 삼각형 - 프로그래머스 Lv. 3
분류 : DP
'''

# 재귀적 구조
def solution(triangle):
    if len(triangle) < 2:
        return triangle[0][0]
    for i in range(len(triangle[-1]) - 1):
        triangle[-2][i] += max(
            triangle[-1][i],
            triangle[-1][i + 1]
        )
    triangle.pop()  # 맨 아랫층 제거
    return solution(triangle)

triangle = [[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]

if __name__ == '__main__':
    print(solution(triangle))

File: Algorithm/2023.11/graywhales/정수삼각형.java
//dp:최적 부분 구조
class Solution {
    int[][] memo;
    int n;
    
    public int solution(int[][] triangle) {
        memo=triangle;
        n=triangle.length;
        
        for(int i=1;i<n;i++){
            memo[i][0]+=memo[i-1][0];//끝부분은 그냥 더하기
            memo[i][i]+=memo[i-1][i-1];
            for(int j=1;j<i;j++){//중간부분 들은 위에 두개중 큰 값 최적 부분 찾기
                memo[i][j]+=Math.max(memo[i-1][j-1],memo[i-1][j]);
            }
        }
        
        int ans=0;
        for(int i=0;i<n;i++){
            ans=Math.max(ans,memo[n-1][i]);
        }
        return ans;
    }
}
